### PR TITLE
[spaceship] Initial support for updated review submissions flow

### PR DESCRIPTION
**Pull request by valerio-castelli**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Apple is introducing a [new submission workflow](https://developer.apple.com/app-store-connect/submission-update/) on App Store Connect. Once an account is updated to follow the new flow, Fastlane's current submission methods will not work. This PR contains the initial work done for adding support for new concepts like `ReviewSubmission`-s and `ReviewSubmissionItem`-s. 

While these endpoints have not been made public from Apple's side, the work done in this PR follows the observed HTTP calls that are made when creating a submission on App Store Connect.

### Description
This PR introduces the concepts of `ReviewSubmission` and `ReviewSubmissionItem`.

Support is added for:
- creating a new review submission
- fetching an existing review submission
- adding items to an existing review submission (only the App Store version, for now)
- submitting a review submission
- cancelling a review submission

The App Store Connect API supports additional concepts, states and operations which have not been included in this PR. In other words, this PR does not aim to provide an exhaustive coverage of the functionalities connected to the new submission flow; rather, its goal is to enable the development of a version of `Deliver` that is compliant with the new submission flow and that can be used when Apple eventually pulls the trigger and enforces it on all developer accounts.

Note that these new classes have not been integrated in `Deliver` yet. This was a conscious choice, as the new submission flow allows users to batch contributions of different kinds (an app version update, a new custom product page, in-app events, etc.); assuming that the submission happens instantaneously as it does now would therefore have been a very opinionated choice, which I think should be taken by the Fastlane community at large.

Rubocop is currently failing due to the increased length of the `App` class caused by the addition of the new submission methods. While I would tentatively argue that such issue could be resolved by re-issuing the rubocop file, I figured it would have been best to hear the community's opinion on the matter.
### Testing Steps

The PR includes unit tests that verify the correctness of the methods and that are executed with the usual `bundle exec rspec` command. In addition, the flow has been tested locally against my company's App Store Connect account using a slightly twisted (and opinionated) version of `Deliver`.

### CODE REVIEW

**Review by joshdholtz — CHANGES_REQUESTED** [review]
This PR is sooo 💯  Really appreciate the tests and fixtures 😍 

It looks like the linter is failing with...
```
spaceship/lib/spaceship/connect_api/models/app.rb:6:5: C: Metrics/ClassLength: Class has too many lines. [342/320]
    class App ...
    ^^^^^^^^^
```

You will need to put `# rubocop:disable Metrics/ClassLength` above the class and `# rubocop:enable Metrics/ClassLength` below the class

You can see an example of this in https://github.com/fastlane/fastlane/blob/master/scan/lib/scan/options.rb

Thanks! ❤️

**Comment by valerio-castelli** [comment]
Thanks for the review, @joshdholtz! I added the two comment lines to the `App` class, the linter should be happy now. 👍

File: spaceship/lib/spaceship/connect_api/models/app.rb [hunk 1]
```diff
@@ -414,6 +415,49 @@ def enable_educational_discount
...
+      end
+
+      # appStoreVersionForReview,items
+      def get_review_submissions(client: nil, filter: {}, includes: nil, limit: nil, sort: nil)
+        client ||= Spaceship::ConnectAPI
+        resps = client.get_review_submissions(app_id: id, filter: filter, includes: includes, limit: limit, sort: sort).all_pages
+        return resps.flat_map(&:to_models)
+      end
+
+      def create_review_submission(client: nil, platform: nil)
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/models/app.rb:455]:
```suggestion
      def create_review_submission(client: nil, platform:)
```

Let's require platform here too 😊  (I need to clean things like this up all over Spaceship 🤦‍♂️ )

File: spaceship/lib/spaceship/connect_api/models/review_submission.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      def self.type
+        return "reviewSubmissions"
+      end
+
+      #
+      # API
+      #
+
+      # appStoreVersionForReview,items,submittedByActor,lastUpdatedByActor
+      def self.get(client: nil, review_submission_id: nil, includes: nil)
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/models/review_submission.rb:46]:
```suggestion
      def self.get(client: nil, review_submission_id:, includes: nil)
```

File: spaceship/lib/spaceship/connect_api/models/review_submission.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      end
+
+      def cancel_submission(client: nil)
+        client ||= Spaceship::ConnectAPI
+        attributes = { canceled: true }
+        resp = client.patch_review_submission(review_submission_id: id, attributes: attributes)
+        return resp.to_models.first
+      end
+
+      def add_app_store_version_to_review_items(client: nil, app_store_version_id: nil)
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/models/review_submission.rb:66]:
```suggestion
      def add_app_store_version_to_review_items(client: nil, app_store_version_id:)
```

File: spaceship/lib/spaceship/connect_api/models/review_submission_item.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      def self.type
+        return "reviewSubmissionItems"
+      end
+
+      #
+      # API
+      #
+
+      # appCustomProductPageVersion,appEvent,appStoreVersion,appStoreVersionExperiment
+      def self.all(client: nil, review_submission_id: nil, includes: nil, limit: nil, sort: nil)
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/models/review_submission_item.rb:33]:
```suggestion
      def self.all(client: nil, review_submission_id:, includes: nil, limit: nil, sort: nil)
```

File: spaceship/lib/spaceship/connect_api/tunes/tunes.rb [hunk 5]
```diff
@@ -1126,6 +1126,89 @@ def delete_idfa_declaration(idfa_declaration_id: nil)
           tunes_request_client.delete("idfaDeclarations/#{idfa_declaration_id}", params)
         end
 
+        #
+        # reviewSubmissions
+        #
+
+        def get_review_submissions(app_id: nil, filter: {}, includes: nil, limit: nil, sort: nil)
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/tunes/tunes.rb:1133]:
```suggestion
        def get_review_submissions(app_id:, filter: {}, includes: nil, limit: nil, sort: nil)
```

File: spaceship/lib/spaceship/connect_api/tunes/tunes.rb [hunk 6]
```diff
@@ -1126,6 +1126,89 @@ def delete_idfa_declaration(idfa_declaration_id: nil)
...
+          params = tunes_request_client.build_params(filter: filter, includes: includes, limit: limit, sort: sort)
+          tunes_request_client.get("apps/#{app_id}/reviewSubmissions", params)
+        end
+
+        def get_review_submission(review_submission_id: nil, filter: {}, includes: nil, limit: nil, sort: nil)
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/tunes/tunes.rb:1138]:
```suggestion
        def get_review_submission(review_submission_id:, filter: {}, includes: nil, limit: nil, sort: nil)
```

File: spaceship/lib/spaceship/connect_api/tunes/tunes.rb [hunk 7]
```diff
@@ -1126,6 +1126,89 @@ def delete_idfa_declaration(idfa_declaration_id: nil)
...
+          params = tunes_request_client.build_params(filter: filter, includes: includes, limit: limit, sort: sort)
+          tunes_request_client.get("reviewSubmissions/#{review_submission_id}", params)
+        end
+
+        def post_review_submission(app_id: nil, platform: nil)
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/tunes/tunes.rb:1143]:
```suggestion
        def post_review_submission(app_id:, platform:)
```

File: spaceship/lib/spaceship/connect_api/tunes/tunes.rb [hunk 8]
```diff
@@ -1126,6 +1126,89 @@ def delete_idfa_declaration(idfa_declaration_id: nil)
...
+                  }
+                }
+              }
+            }
+          }
+
+          tunes_request_client.post("reviewSubmissions", body)
+        end
+
+        def patch_review_submission(review_submission_id: nil, attributes: nil)
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/tunes/tunes.rb:1164]:
```suggestion
        def patch_review_submission(review_submission_id:, attributes: nil)
```

File: spaceship/lib/spaceship/connect_api/tunes/tunes.rb [hunk 9]
```diff
@@ -1126,6 +1126,89 @@ def delete_idfa_declaration(idfa_declaration_id: nil)
...
+          }
+
+          tunes_request_client.patch("reviewSubmissions/#{review_submission_id}", body)
+        end
+
+        #
+        # reviewSubmissionItems
+        #
+
+        def get_review_submission_items(review_submission_id: nil, filter: {}, includes: nil, limit: nil, sort: nil)
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/tunes/tunes.rb:1180]:
```suggestion
        def get_review_submission_items(review_submission_id:, filter: {}, includes: nil, limit: nil, sort: nil)
```

File: spaceship/lib/spaceship/connect_api/tunes/tunes.rb [hunk 10]
```diff
@@ -1126,6 +1126,89 @@ def delete_idfa_declaration(idfa_declaration_id: nil)
...
+          params = tunes_request_client.build_params(filter: filter, includes: includes, limit: limit, sort: sort)
+          tunes_request_client.get("reviewSubmissions/#{review_submission_id}/items", params)
+        end
+
+        def post_review_submission_item(review_submission_id: nil, app_store_version_id: nil)
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/tunes/tunes.rb:1185]:
```suggestion
        def post_review_submission_item(review_submission_id:, app_store_version_id: nil)
```

File: spaceship/lib/spaceship/connect_api/models/app.rb [hunk 11]
```diff
@@ -414,6 +415,49 @@ def enable_educational_discount
         })
       end
 
+      #
+      # Review Submissions
+      #
+
+      def get_ready_review_submission(client: nil, platform: nil, includes: nil)
+        client ||= Spaceship::ConnectAPI
+        platform ||= Spaceship::ConnectAPI::Platform::IOS
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/models/app.rb:424]:
```suggestion
      def get_ready_review_submission(client: nil, platform:, includes: nil)
        client ||= Spaceship::ConnectAPI
```

Thoughts on requiring platform and removing the default? It got me in trouble and confused when testing this on a Mac app 😛 I know we do this in other places in spaceship but I think I'd like to get away from this if possible 🤷‍♂️

File: spaceship/lib/spaceship/connect_api/models/app.rb [hunk 12]
```diff
@@ -414,6 +415,49 @@ def enable_educational_discount
...
+          ].join(","),
+          platform: platform
+        }
+
+        return get_review_submissions(client: client, filter: filter, includes: includes).first
+      end
+
+      def get_in_progress_review_submission(client: nil, platform: nil, includes: nil)
+        client ||= Spaceship::ConnectAPI
+        platform ||= Spaceship::ConnectAPI::Platform::IOS
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/models/app.rb:437]:
```suggestion
      def get_in_progress_review_submission(client: nil, platform:, includes: nil)
        client ||= Spaceship::ConnectAPI
```

Same as above

**Review by joshdholtz — CHANGES_REQUESTED** [review]
This is 💯 But also like more than that so its like 💯 💯 💯  Thank you so much for adding this but also adding tests 😊 

I left a handful of small comments on changing the named arguments for methods but other than this looks good! I didn't test all of the functionality yet but I did create a submission, added a version, submitted it, and cancelled it and it worked great!

File: spaceship/lib/spaceship/connect_api/models/app.rb (hunk 11 above)
Comment (valerio-castelli) [reply spaceship/lib/spaceship/connect_api/models/app.rb]: I've accepted all your suggestions for mandatory fields! 😀

**Review by joshdholtz — APPROVED** [review]
This looks great! Thank you sooooo much for adding this 😍 Really really love the tests 🚀

**Comment by valerio-castelli** [comment]
@joshdholtz Just a heads up that Apple plans to start enforcing the new submission flow from January 25th, so `Deliver` will most likely start to fail (at least for some people) unless we integrate this new logic. This is what they say on App Store Connect:
```
Last year, we introduced an updated way to submit apps, in-app events, custom product pages, and product page optimization tests for review. Starting January 25, 2022, the submission experience will be automatically updated for all App Store Connect accounts.
```

I don't know what the best approach would be to support this within `Deliver`, especially if we consider that the new submission flow makes it possible to have "asynchronous" or "deferred" submissions (you can create the submission, add items to it, and submit the items in bulk) – which may be against the way `Deliver` has traditionally operated.

I hope that by letting you know in advance (if you're not already aware of it) we may be able to smoothen the transition when January 25th comes 🙂

**Comment by joshdholtz** [comment]
@valerio-castelli I have this started in https://github.com/fastlane/fastlane/pull/19838 😊 I'm going to be finishing it up tomorrow and then validate it whenever the App Store Connect API gets the official support for this 💪

**Pull request merged**